### PR TITLE
Fix dialyzer warnings

### PR DESCRIPTION
**Pull request by rustra**
PR adds missing type specs and ignores some Dialyzer warnings.

### CODE REVIEW

File: lib/rdf/blank_node/increment.ex [hunk 1]
```diff
@@ -31,6 +31,7 @@ defmodule RDF.BlankNode.Increment do
   end
 
   @impl BlankNode.Generator.Algorithm
+  # @spec generate(map) :: {RDF.BlankNode.t, map}
```
[the remark this answers is not in the document]
Comment (rustra) [reply lib/rdf/blank_node/increment.ex]: Sure, fixed.

File: lib/rdf/dataset.ex [hunk 2]
```diff
@@ -13,21 +13,30 @@ defmodule RDF.Dataset do
...
-  defstruct name: nil, graphs: %{}
-
   @behaviour Access
 
-  alias RDF.{Graph, Description}
+  alias RDF.{Description, Graph, IRI, Statement}
   import RDF.Statement
 
-  @type t :: module
+  @type graphs_key :: IRI.t | nil
```
[the remark this answers is not in the document]
Comment (rustra) [reply lib/rdf/dataset.ex]: No problem, changed!

File: lib/rdf/dataset.ex [hunk 3]
```diff
@@ -13,21 +13,30 @@ defmodule RDF.Dataset do
...
+
+  @type t :: %__MODULE__{
+          name: IRI.t | nil,
```
[the remark this answers is not in the document]
Comment (rustra) [reply lib/rdf/dataset.ex]: Thanks, fixed.

File: lib/rdf/iri.ex [hunk 4]
```diff
@@ -16,12 +16,16 @@ defmodule RDF.IRI do
...
-  @enforce_keys [:value]
-  defstruct [:value]
+  alias RDF.{Datatype, Namespace}
+
+  @type t :: %__MODULE__{
+          value: String.t
+  }
 
-  alias RDF.Namespace
+  @type t_param :: String.t | Datatype.t | URI.t | t
```
[the remark this answers is not in the document]
Comment (rustra) [reply lib/rdf/iri.ex]: Thank you a lot for clarifications! I've changed it.

File: lib/rdf/iri.ex [hunk 5]
```diff
@@ -245,6 +260,7 @@ defmodule RDF.IRI do
       "http://example.com/#Foo"
 
   """
+  @spec to_string(t | Datatype.t) :: String.t
```
[the remark this answers is not in the document]
Comment (rustra) [reply lib/rdf/iri.ex]: Done.

File: lib/rdf/prefix_map.ex [hunk 6]
```diff
@@ -5,14 +5,27 @@ defmodule RDF.PrefixMap do
   `RDF.PrefixMap` implements the `Enumerable` protocol.
   """
 
+  alias RDF.IRI
+
+  @type prefix    :: atom | String.t
+  @type namespace :: atom | String.t | IRI.t
```
[the remark this answers is not in the document]
Comment (rustra) [reply lib/rdf/prefix_map.ex]: Thanks, changed it.

File: lib/rdf/graph.ex [hunk 7]
```diff
@@ -11,18 +11,33 @@ defmodule RDF.Graph do
...
 
-  defstruct name: nil, descriptions: %{}, prefixes: nil, base_iri: nil
-
   @behaviour Access
 
-  alias RDF.Description
   import RDF.Statement
+  alias RDF.{Description, IRI, PrefixMap, Statement}
+
+  @type graph_description :: %{IRI.t => Description.t}
```
[the remark this answers is not in the document]
Comment (rustra) [reply lib/rdf/graph.ex]: Done.

File: lib/rdf/graph.ex [hunk 8]
```diff
@@ -11,18 +11,33 @@ defmodule RDF.Graph do
...
+
+  @type t :: %__MODULE__{
+          name: IRI.t | nil,
+          descriptions: graph_description,
+          prefixes: PrefixMap.t | nil,
+          base_iri: IRI.t | nil
+  }
+
+  @type t_param :: Statement.t | Description.t | t
```
[the remark this answers is not in the document]
Comment (rustra) [reply lib/rdf/graph.ex]: I think that it's better name )

File: lib/rdf/dataset.ex [hunk 9]
```diff
@@ -13,21 +13,30 @@ defmodule RDF.Dataset do
...
+          graphs: %{graphs_key => Graph.t}
+  }
+
+  @type t_param :: Graph.t_param | t
```
[the remark this answers is not in the document]
Comment (rustra) [reply lib/rdf/dataset.ex]: Agree!

File: lib/rdf/literal.ex [hunk 10]
```diff
@@ -345,12 +381,13 @@ defmodule RDF.Literal do
         |> String.downcase()
         |> String.contains?(String.downcase(pattern))
 
-      _ ->
-        raise "Invalid XQuery regex pattern or flags"
+      # _ ->
+      #   raise "Invalid XQuery regex pattern or flags"
```
[the remark this answers is not in the document]
Comment (rustra) [reply lib/rdf/literal.ex]: Good point. I missed that `Regex.compile` could return `{:error, any}`, so in that case `raise` could be necessary. Fixed.

File: lib/rdf/prefix_map.ex [hunk 11]
```diff
@@ -7,12 +7,16 @@ defmodule RDF.PrefixMap do
...
-  @type namespace :: atom | String.t | IRI.t
+  @type prefix    :: atom
+  @type namespace :: IRI.t
 
-  @type prefix_map :: %{prefix => namespace}
+  @type coercible_prefix    :: atom | String.t
+  @type coercible_namespace :: atom | String.t | IRI.t
 
-  @type conflict_resolver :: (prefix, namespace, namespace -> namespace)
+  @type prefix_map :: %{coercible_prefix => coercible_namespace}
```
Comment (marcelotto) [inline lib/rdf/prefix_map.ex]: In this type only the strict types should be allowed

```suggestion
  @type prefix_map :: %{prefix => namespace}
```

File: lib/rdf/datatypes/numeric.ex [hunk 12]
```diff
@@ -35,11 +35,13 @@ defmodule RDF.Numeric do
   @doc """
   The list of all numeric datatypes.
   """
+  @dialyzer {:nowarn_function, types: 0}
```
[the remark this answers is not in the document]
Comment (marcelotto) [reply lib/rdf/datatypes/numeric.ex]: What about this? Doesn't it work with the suggested spec? I could live without it but I'm curious.

**Comment by rustra** [comment]
According this: https://github.com/marcelotto/rdf-ex/pull/6#discussion_r387253064
Actually the problem is that I don't really understand how to deal with MapSet as it has `@type t` and `@opaque t(value)`.
Dialyzer says: `The call 'Elixir.MapSet':to_list(...) does not have an opaque term of type 'Elixir.MapSet':t(_) as 1st argument`.
I believe that some beautiful solution should exist and somebody else will fix it later.

File: lib/rdf/prefix_map.ex (hunk 11 above)
Comment (rustra) [reply lib/rdf/prefix_map.ex]: Thanks, fixed.

File: lib/rdf/literal.ex [hunk 13]
```diff
@@ -3,11 +3,30 @@ defmodule RDF.Literal do
...
+          | integer
+          | float
+          | String.t
+          | Decimal.t
+          | Date.t
+          | Time.t
+          | DateTime.t
+          | NaiveDateTime.t
+          | {Date.t, String.t}
+          | :nan
```
[the remark this answers is not in the document]
Comment (rustra) [reply lib/rdf/literal.ex]: It sounds reasonable. Added your proposal.

File: lib/rdf/datatypes/numeric.ex [hunk 14]
```diff
@@ -35,11 +35,13 @@ defmodule RDF.Numeric do
...
   def types(), do: MapSet.to_list(@types)
 
   @doc """
   Returns if a given datatype is a numeric datatype.
   """
+  @dialyzer {:nowarn_function, type?: 1}
```
[the remark this answers is not in the document]
Comment (rustra) [reply lib/rdf/datatypes/numeric.ex]: https://github.com/marcelotto/rdf-ex/pull/6#issuecomment-594166992

File: lib/rdf/datatypes/numeric.ex (hunk 12 above)
[the remark this answers is not in the document]
Comment (rustra) [reply lib/rdf/datatypes/numeric.ex]: https://github.com/marcelotto/rdf-ex/pull/6#issuecomment-594166992

**Pull request merged**